File: src/bus/canonical_scope.cpp
#include "starling/bus/canonical_scope.hpp"

#include "starling/extractor/extracted_statement.hpp"
#include "starling/schema/statement_enums.hpp"

#include <algorithm>

namespace starling::bus {

std::string CanonicalScopeNorm::canonical_bytes() const {
    std::string out = kind;
    for (const auto& m : members_sorted) { out += '\x1f'; out += m; }
    return out;
}

std::string CanonicalScopeCommitment::canonical_bytes() const {
    return principal + '\x1f' + beneficiary;
}

std::string CanonicalScopeCommonGround::canonical_bytes() const {
    std::string out;
    bool first = true;
    for (const auto& p : parties_sorted) { if (!first) out += '\x1f'; out += p; first = false; }
    return out;
}

std::string canonical_scope_bytes(const CanonicalScope& scope) {
    return std::visit([](const auto& arm) { return arm.canonical_bytes(); }, scope);
}

CanonicalScope scope_of(const starling::extractor::ExtractedStatement& stmt) {
    using M = schema::Modality;
    if (stmt.modality == M::COMMITS) {
        std::string beneficiary;
        for (const auto& p : stmt.scope_parties) {
            if (p != stmt.holder_id) { beneficiary = p; break; }
        }
        return CanonicalScopeCommitment{stmt.holder_id, beneficiary};
    }
    if (stmt.modality == M::NORM_OUGHT || stmt.modality == M::NORM_FORBID) {
        std::vector<std::string> members{stmt.holder_id, stmt.subject_id};
        std::sort(members.begin(), members.end());
        const std::string kind = (stmt.modality == M::NORM_OUGHT) ? "obligation" : "prohibition";
        return CanonicalScopeNorm{kind, members};
    }
    if (stmt.scope_parties.size() >= 2) {
        std::vector<std::string> parties = stmt.scope_parties;
        std::sort(parties.begin(), parties.end());
        return CanonicalScopeCommonGround{parties};
    }
    return CanonicalScopeNull{};
}

}  // namespace starling::bus

File: python/starling/bus/canonical_scope.py
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Any, List, Union


@dataclass(frozen=True)
class CanonicalScopeNull:
    def canonical_bytes(self) -> str:
        return ""


@dataclass(frozen=True)
class CanonicalScopeNorm:
    kind: str = ""
    members_sorted: tuple = ()

    def canonical_bytes(self) -> str:
        return self.kind + ''.join('\x1f' + m for m in self.members_sorted)


@dataclass(frozen=True)
class CanonicalScopeCommitment:
    principal: str = ""
    beneficiary: str = ""

    def canonical_bytes(self) -> str:
        return self.principal + '\x1f' + self.beneficiary


@dataclass(frozen=True)
class CanonicalScopeCommonGround:
    parties_sorted: tuple = ()

    def canonical_bytes(self) -> str:
        return '\x1f'.join(self.parties_sorted)


CanonicalScope = Union[
    CanonicalScopeNull,
    CanonicalScopeNorm,
    CanonicalScopeCommitment,
    CanonicalScopeCommonGround,
]


def canonical_scope_bytes(scope: CanonicalScope) -> str:
    return scope.canonical_bytes()


def scope_of(stmt: Any) -> CanonicalScope:
    """Derive the canonical scope from a statement (P2.j).

    Mirrors the C++ scope_of byte-for-byte: modality 优先（COMMITS→Commitment；
    NORM_*→Norm；else scope_parties≥2→CommonGround；else Null）。
    """
    mod = stmt.modality.name if hasattr(stmt.modality, 'name') else str(stmt.modality)
    parties = list(getattr(stmt, 'scope_parties', []) or [])
    holder = stmt.holder_id
    if mod == 'COMMITS':
        beneficiary = next((p for p in parties if p != holder), "")
        return CanonicalScopeCommitment(principal=holder, beneficiary=beneficiary)
    if mod in ('NORM_OUGHT', 'NORM_FORBID'):
        members = sorted([holder, stmt.subject_id])
        kind = 'obligation' if mod == 'NORM_OUGHT' else 'prohibition'
        return CanonicalScopeNorm(kind=kind, members_sorted=members)
    if len(parties) >= 2:
        return CanonicalScopeCommonGround(parties_sorted=sorted(parties))
    return CanonicalScopeNull()
